# MRETT Results

Complete spreadsheet: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1384530880 

MRETT21: 1gE2gzncRuiwzY4xgKEbVDjSBUde8LGc7u9pIMQ0C6bQ 

## Importing Google Spreadsheet into Pandas

https://stackoverflow.com/questions/19611729/getting-google-spreadsheet-csv-into-a-pandas-dataframe

```python
pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                   '0Ak1ecr7i0wotdGJmTURJRnZLYlV3M2daNTRubTdwTXc' +
                   '/export?gid=0&format=csv',
                   # Set first column as rownames in data frame
                   index_col=0,
                   # Parse column values to datetime
                   parse_dates=['Quradate']
                  )
```

## APIs

https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/values/MRETT22

https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit#gid=171816323&range=C17

GC: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1384530880
21: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=0
22: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=601708178
23: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=171816323
24: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1769801856
25: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1953569581
26: https://docs.google.com/spreadsheets/d/11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU/edit?fbclid=IwAR2KA8_pks7JY4Q_P7XLrNSawnmysBJo2FnAFvjUe5r7XiAlAi134Z-PXNw#gid=1178762166

In [399]:
import pandas as pd

df_MRETT_GC = pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                   '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' +
                   '/export?gid=' + '1384530880' + '&format=csv',
                   # Set first column as rownames in data frame
                   index_col=0,
                   # Parse column values to datetime
                   #parse_dates=['Quradate']
                  )

#df_MRETT22 = pd.read_csv('https://docs.google.com/spreadsheets/d/' + '1gIz_Gs6uxUrMUJKk7SrflaAiKeOKScQfvNqnX4bADV0' + '/export?gid=' + '1667031214' + '&format=csv', index_col=0)

df_MRETT_GC


Team  Total MRETT21 MRETT22 MRETT23 MRETT24 MRETT25  \
#                                                                          
1              Pocomotion    875             165     170     165     175   
2                  eCKD 1    800             200     200     200     200   
2                OTR Blue    800             170     145     145     150   
4   Team Lou Racing Squad    775             145     150     155     145   
5            The Pedalers    750             190     190     190     180   
6            Rasio Racing    730             180     180     180     190   
7               OTR Black    670             160     175     175     160   
8      KISS Racing Team 1    665             175     155     170     165   
9             Moon Riders    631             141     160     160     170   
10              OTR Green    607             137     165     150     155   
11                 eCKD 2    548             133     137     137     141   
12    Westerley CC Purple    540             129     133     141     137   
13             Westerlies    504             121     129     125     129   
14           Team Lou Too    441     NaN     NaN     NaN     133     133   
15          WKG Renegades    395             125     141     129           
16           RGT France 1    155             155                           
17              Les Watts    150             150                           
18        Tricky Allsorts      0                                           

   MRETT26  
#           
1      200  
2           
2      190  
4      180  
5           
6           
7           
8           
9           
10          
11          
12          
13          
14     175  
15          
16          
17          
18

## Wait for Spreadsheet to Initialize

Since the spreadsheet reads data from other sheets, it may take several seconds for it to properly fetch and calculate all values.

In [400]:
import time

for i in range(5):
    time.sleep(5) 
    pd.read_csv('https://docs.google.com/spreadsheets/d/' + 
                    '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' +
                    '/export?' + '' + '&format=csv',
                    index_col=1,
                    )

## Get all Sheet Tabs, Including GC

In [445]:
mrett_gids = {
    'GC' : '1384530880',
    '21' : '0',
    '22' : '601708178',
    '23' : '171816323',
    '24' : '1769801856',
    '25' : '1953569581',
    '26' : '1178762166'
}

df_mretts = {}

for mrett in mrett_gids.keys():
    print(mrett, mrett_gids[mrett])
    gid = mrett_gids[mrett]
    df_mretts[mrett] = pd.read_csv('https://docs.google.com/spreadsheets/d/' + '11uk0WpQzjVBOWs99cev9buajG1PalFdC5mtdynhqejU' + '/export?gid=' + gid + '&format=csv', index_col=1, na_values=['', ' '])

df_mretts['GC']

GC 1384530880
21 0
22 601708178
23 171816323
24 1769801856
25 1953569581
26 1178762166


#  Total  MRETT21  MRETT22  MRETT23  MRETT24  MRETT25  \
Team                                                                            
Pocomotion              1    875      NaN    165.0    170.0    165.0    175.0   
eCKD 1                  2    800      NaN    200.0    200.0    200.0    200.0   
OTR Blue                2    800      NaN    170.0    145.0    145.0    150.0   
Team Lou Racing Squad   4    775      NaN    145.0    150.0    155.0    145.0   
The Pedalers            5    750      NaN    190.0    190.0    190.0    180.0   
Rasio Racing            6    730      NaN    180.0    180.0    180.0    190.0   
OTR Black               7    670      NaN    160.0    175.0    175.0    160.0   
KISS Racing Team 1      8    665      NaN    175.0    155.0    170.0    165.0   
Moon Riders             9    631      NaN    141.0    160.0    160.0    170.0   
OTR Green              10    607      NaN    137.0    165.0    150.0    155.0   
eCKD 2                 11    548      NaN    133.0    137.0    137.0    141.0   
Westerley CC Purple    12    540      NaN    129.0    133.0    141.0    137.0   
Westerlies             13    504      NaN    121.0    129.0    125.0    129.0   
Team Lou Too           14    441      NaN      NaN      NaN    133.0    133.0   
WKG Renegades          15    395      NaN    125.0    141.0    129.0      NaN   
RGT France 1           16    155      NaN    155.0      NaN      NaN      NaN   
Les Watts              17    150      NaN    150.0      NaN      NaN      NaN   
Tricky Allsorts        18      0      NaN      NaN      NaN      NaN      NaN   

                       MRETT26  
Team                            
Pocomotion               200.0  
eCKD 1                     NaN  
OTR Blue                 190.0  
Team Lou Racing Squad    180.0  
The Pedalers               NaN  
Rasio Racing               NaN  
OTR Black                  NaN  
KISS Racing Team 1         NaN  
Moon Riders                NaN  
OTR Green                  NaN  
eCKD 2                     NaN  
Westerley CC Purple        NaN  
Westerlies                 NaN  
Team Lou Too             175.0  
WKG Renegades              NaN  
RGT France 1               NaN  
Les Watts                  NaN  
Tricky Allsorts            NaN

In [446]:
if pd.isna(df_mretts['GC']['MRETT21'].iloc[0]):
    df_mretts['GC'].at['eCKD 1', 'MRETT21'] = 200
    df_mretts['GC'].at['Rasio Racing', 'MRETT21'] = 190
    df_mretts['GC'].at['The Pedalers', 'MRETT21'] = 170
    df_mretts['GC'].at['Pocomotion', 'MRETT21'] = 175
    df_mretts['GC'].at['OTR Black', 'MRETT21'] = 165
    df_mretts['GC'].at['KISS Racing Team 1', 'MRETT21'] = 150
    df_mretts['GC'].at['Moon Riders', 'MRETT21'] = 160
    df_mretts['GC'].at['OTR Blue', 'MRETT21'] = 155
    df_mretts['GC'].at['OTR Green', 'MRETT21'] = 141
    df_mretts['GC'].at['Team Lou Racing Squad', 'MRETT21'] = 137
    df_mretts['GC'].at['eCKD 2', 'MRETT21'] = 145
    df_mretts['GC'].at['Westerley CC Purple', 'MRETT21'] = 129
    df_mretts['GC'].at['WKG Renegades', 'MRETT21'] = 125
    df_mretts['GC'].at['Westerlies', 'MRETT21'] = 0
    df_mretts['GC'].at['RGT France 1', 'MRETT21'] = 189
    df_mretts['GC'].at['Team Lou Too', 'MRETT21'] = 0
    df_mretts['GC'].at['Les Watts', 'MRETT21'] = 0
    df_mretts['GC'].at['Tricky Allsorts', 'MRETT21'] = 133
    df_mretts['GC']['Total']  = df_mretts['GC']['Total'] + df_mretts['GC']['MRETT21'] 


df_mretts['GC']

#   Total  MRETT21  MRETT22  MRETT23  MRETT24  \
Team                                                                    
Pocomotion              1  1050.0    175.0    165.0    170.0    165.0   
eCKD 1                  2  1000.0    200.0    200.0    200.0    200.0   
OTR Blue                2   955.0    155.0    170.0    145.0    145.0   
Team Lou Racing Squad   4   912.0    137.0    145.0    150.0    155.0   
The Pedalers            5   920.0    170.0    190.0    190.0    190.0   
Rasio Racing            6   920.0    190.0    180.0    180.0    180.0   
OTR Black               7   835.0    165.0    160.0    175.0    175.0   
KISS Racing Team 1      8   815.0    150.0    175.0    155.0    170.0   
Moon Riders             9   791.0    160.0    141.0    160.0    160.0   
OTR Green              10   748.0    141.0    137.0    165.0    150.0   
eCKD 2                 11   693.0    145.0    133.0    137.0    137.0   
Westerley CC Purple    12   669.0    129.0    129.0    133.0    141.0   
Westerlies             13   504.0      0.0    121.0    129.0    125.0   
Team Lou Too           14   441.0      0.0      NaN      NaN    133.0   
WKG Renegades          15   520.0    125.0    125.0    141.0    129.0   
RGT France 1           16   344.0    189.0    155.0      NaN      NaN   
Les Watts              17   150.0      0.0    150.0      NaN      NaN   
Tricky Allsorts        18   133.0    133.0      NaN      NaN      NaN   

                       MRETT25  MRETT26  
Team                                     
Pocomotion               175.0    200.0  
eCKD 1                   200.0      NaN  
OTR Blue                 150.0    190.0  
Team Lou Racing Squad    145.0    180.0  
The Pedalers             180.0      NaN  
Rasio Racing             190.0      NaN  
OTR Black                160.0      NaN  
KISS Racing Team 1       165.0      NaN  
Moon Riders              170.0      NaN  
OTR Green                155.0      NaN  
eCKD 2                   141.0      NaN  
Westerley CC Purple      137.0      NaN  
Westerlies               129.0      NaN  
Team Lou Too             133.0    175.0  
WKG Renegades              NaN      NaN  
RGT France 1               NaN      NaN  
Les Watts                  NaN      NaN  
Tricky Allsorts            NaN      NaN

In [444]:

""" df_mretts['GC'].at['eCKD 1', 'MRETT21'] = 200
df_mretts['GC'].at['Rasio Racing', 'MRETT21'] = 190
df_mretts['GC'].at['The Pedalers', 'MRETT21'] = 170
df_mretts['GC'].at['Pocomotion', 'MRETT21'] = 175
df_mretts['GC'].at['OTR Black', 'MRETT21'] = 165
df_mretts['GC'].at['KISS Racing Team 1', 'MRETT21'] = 150
df_mretts['GC'].at['Moon Riders', 'MRETT21'] = 160
df_mretts['GC'].at['OTR Blue', 'MRETT21'] = 155
df_mretts['GC'].at['OTR Green', 'MRETT21'] = 141
df_mretts['GC'].at['Team Lou Racing Squad', 'MRETT21'] = 137
df_mretts['GC'].at['eCKD 2', 'MRETT21'] = 145
df_mretts['GC'].at['Westerley CC Purple', 'MRETT21'] = 129
df_mretts['GC'].at['WKG Renegades', 'MRETT21'] = 125
df_mretts['GC'].at['Westerlies', 'MRETT21'] = 0
df_mretts['GC'].at['RGT France 1', 'MRETT21'] = 189
df_mretts['GC'].at['Team Lou Too', 'MRETT21'] = 0
df_mretts['GC'].at['Les Watts', 'MRETT21'] = 0
df_mretts['GC'].at['Tricky Allsorts', 'MRETT21'] = 133

df_mretts['GC'] """


#  Total  MRETT21  MRETT22  MRETT23  MRETT24  MRETT25  \
Team                                                                            
Pocomotion              1   1075      175    165.0    170.0    165.0    175.0   
eCKD 1                  2    990      200    200.0    200.0    200.0    200.0   
OTR Blue                2    970      155    170.0    145.0    145.0    150.0   
Team Lou Racing Squad   4    950      137    145.0    150.0    155.0    145.0   
The Pedalers            5    915      170    190.0    190.0    190.0    180.0   
Rasio Racing            6    880      190    180.0    180.0    180.0    190.0   
OTR Black               7    830      165    160.0    175.0    175.0    160.0   
KISS Racing Team 1      8    820      150    175.0    155.0    170.0    165.0   
Moon Riders             9    772      160    141.0    160.0    160.0    170.0   
OTR Green              10    744      141    137.0    165.0    150.0    155.0   
eCKD 2                 11    693      145    133.0    137.0    137.0    141.0   
Westerley CC Purple    12    669      129    129.0    133.0    141.0    137.0   
Westerlies             13    629        0    121.0    129.0    125.0    129.0   
Team Lou Too           14    441        0      NaN      NaN    133.0    133.0   
WKG Renegades          15    584      125    125.0    141.0    129.0      NaN   
RGT France 1           16    155      189    155.0      NaN      NaN      NaN   
Les Watts              17    150        0    150.0      NaN      NaN      NaN   
Tricky Allsorts        18    133      133      NaN      NaN      NaN      NaN   

                       MRETT26  
Team                            
Pocomotion               200.0  
eCKD 1                     NaN  
OTR Blue                 190.0  
Team Lou Racing Squad    180.0  
The Pedalers               NaN  
Rasio Racing               NaN  
OTR Black                  NaN  
KISS Racing Team 1         NaN  
Moon Riders                NaN  
OTR Green                  NaN  
eCKD 2                     NaN  
Westerley CC Purple        NaN  
Westerlies                 NaN  
Team Lou Too             175.0  
WKG Renegades              NaN  
RGT France 1               NaN  
Les Watts                  NaN  
Tricky Allsorts            NaN

	#	Total	MRETT21	MRETT22	MRETT23	MRETT24	MRETT25	MRETT26
Team								
eCKD 1	1	1000	200.0	200.0	200.0	200.0	200.0	NaN
Rasio Racing	2	920	190.0	180.0	180.0	180.0	190.0	NaN
The Pedalers	2	920	170.0	190.0	190.0	190.0	180.0	NaN
Pocomotion	4	850	175.0	165.0	170.0	165.0	175.0	NaN
OTR Black	5	835	165.0	160.0	175.0	175.0	160.0	NaN
KISS Racing Team 1	6	815	150.0	175.0	155.0	170.0	165.0	NaN
Moon Riders	7	791	160.0	141.0	160.0	160.0	170.0	NaN
OTR Blue	8	765	155.0	170.0	145.0	145.0	150.0	NaN
OTR Green	9	748	141.0	137.0	165.0	150.0	155.0	NaN
Team Lou Racing Squad	10	732	137.0	145.0	150.0	155.0	145.0	NaN
eCKD 2	11	693	145.0	133.0	137.0	137.0	141.0	NaN
Westerley CC Purple	12	669	129.0	129.0	133.0	141.0	137.0	NaN
WKG Renegades	13	520	125.0	125.0	141.0	129.0	NaN	NaN
Westerlies	14	504	NaN	121.0	129.0	125.0	129.0	NaN
RGT France 1	15	335	180.0	155.0	NaN	NaN	NaN	NaN
Team Lou Too	16	266	NaN	NaN	NaN	133.0	133.0	NaN
Les Watts	17	150	NaN	150.0	NaN	NaN	NaN	NaN
Tricky Allsorts	18	133	133.0	NaN	NaN	NaN	NaN	NaN


In [402]:
import pandas_bokeh
import bokeh
from bokeh.themes import built_in_themes
from bokeh.io import curdoc

pandas_bokeh.output_notebook()


Loading BokehJS ...

In [447]:

df_mretts['GC'][::-1].loc[:,'MRETT21':'MRETT26'].fillna(0).plot_bokeh.barh(
    ylabel="Team",
    xlabel="Points", 
    title="MRETT Series 2 GC", 
    stacked=True,
    alpha=0.6,
    figsize=(1600, 600),
    legend = "bottom_right",
    sizing_mode="scale_width"
    )

Figure(id='41442', ...)

## Convert Time column to timedelta

In [449]:
for mrett in mrett_gids.keys():
    if mrett != 'GC':
        df_mretts[mrett]['timedelta'] = pd.to_timedelta(df_mretts[mrett].Time) 
        df_mretts[mrett]['Time in Seconds'] = df_mretts[mrett]['timedelta'].dt.total_seconds() 
        df_mretts[mrett]['Time in Minutes'] = df_mretts[mrett]['Time in Seconds']/60
        df_mretts[mrett]['Time in Minutes Truncated'] = divmod(df_mretts[mrett]['Time in Seconds'], 60)[0] 
        df_mretts[mrett]['Time Remainder in Seconds'] = divmod(df_mretts[mrett]['Time in Seconds'], 60)[1] 
        df_mretts[mrett]['Team Name'] = df_mretts[mrett].index
        df_mretts[mrett]
df_mretts['21']

#  Time  Diff  Points timedelta  Time in Seconds  \
Team                                                                       
eCKD 1                NaN   NaN   NaN     NaN       NaT              NaN   
eCKD 2                NaN   NaN   NaN     NaN       NaT              NaN   
KISS Racing Team 1    NaN   NaN   NaN     NaN       NaT              NaN   
Moon Riders           NaN   NaN   NaN     NaN       NaT              NaN   
OTR Black             NaN   NaN   NaN     NaN       NaT              NaN   
OTR Blue              NaN   NaN   NaN     NaN       NaT              NaN   
OTR Green             NaN   NaN   NaN     NaN       NaT              NaN   
Pocomotion            NaN   NaN   NaN     NaN       NaT              NaN   
Rasio Racing          NaN   NaN   NaN     NaN       NaT              NaN   
RGT France 1          NaN   NaN   NaN     NaN       NaT              NaN   
Team Lou Racing Squad NaN   NaN   NaN     NaN       NaT              NaN   
The Pedalers          NaN   NaN   NaN     NaN       NaT              NaN   
Tricky Allsorts       NaN   NaN   NaN     NaN       NaT              NaN   
Westerley CC Purple   NaN   NaN   NaN     NaN       NaT              NaN   
WKG Renegades         NaN   NaN   NaN     NaN       NaT              NaN   

                       Time in Minutes  Time in Minutes Truncated  \
Team                                                                
eCKD 1                             NaN                        NaN   
eCKD 2                             NaN                        NaN   
KISS Racing Team 1                 NaN                        NaN   
Moon Riders                        NaN                        NaN   
OTR Black                          NaN                        NaN   
OTR Blue                           NaN                        NaN   
OTR Green                          NaN                        NaN   
Pocomotion                         NaN                        NaN   
Rasio Racing                       NaN                        NaN   
RGT France 1                       NaN                        NaN   
Team Lou Racing Squad              NaN                        NaN   
The Pedalers                       NaN                        NaN   
Tricky Allsorts                    NaN                        NaN   
Westerley CC Purple                NaN                        NaN   
WKG Renegades                      NaN                        NaN   

                       Time Remainder in Seconds              Team Name  
Team                                                                     
eCKD 1                                       NaN                 eCKD 1  
eCKD 2                                       NaN                 eCKD 2  
KISS Racing Team 1                           NaN     KISS Racing Team 1  
Moon Riders                                  NaN            Moon Riders  
OTR Black                                    NaN              OTR Black  
OTR Blue                                     NaN               OTR Blue  
OTR Green                                    NaN              OTR Green  
Pocomotion                                   NaN             Pocomotion  
Rasio Racing                                 NaN           Rasio Racing  
RGT France 1                                 NaN           RGT France 1  
Team Lou Racing Squad                        NaN  Team Lou Racing Squad  
The Pedalers                                 NaN           The Pedalers  
Tricky Allsorts                              NaN        Tricky Allsorts  
Westerley CC Purple                          NaN    Westerley CC Purple  
WKG Renegades                                NaN          WKG Renegades

In [450]:
df_mretts['21'][::-1].plot_bokeh.barh(    
    title = 'MRETT' + '21',
    xlabel = 'Time (Minutes)',
    y='Time in Minutes',
    alpha=0.6,
    figsize=(1600, 300),
    legend = "top_right",
    #hovertool_columns='Time',,
    hovertool_string="""<h2> #@{#}                             
                        @{Team Name} </h2>
                        <h3> Time: @{Time} </h3>""",
    sizing_mode="scale_width")

Figure(id='41735', ...)

In [451]:
for mrett in mrett_gids.keys():
    if mrett != 'GC':
        df_mretts[mrett][::-1].plot_bokeh.barh(    
            title = 'MRETT' + mrett,
            xlabel = 'Time (Minutes)',
            y='Time in Minutes',
            alpha=0.6,
            figsize=(1600, 300),
            legend = "top_right",
            hovertool_string="""<h2> #@{#} 
                                @{Team Name} </h2>
                                <h3> Time: @{Time} </h3>""",
            sizing_mode="scale_width")
